main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -505,16 +518,27 @@ object Lowering {
       val e = visitExp(exp)
       LoweredAst.Expr.ApplyAtomic(AtomicOp.ArrayLength, List(e), Type.Int32, eff, loc)
 
-    case TypedAst.Expr.StructNew(sym, fields, region, tpe, eff, loc) => throw new RuntimeException("JOE TBD")
-    case TypedAst.Expr.StructGet(sym, exp, field, tpe, eff, loc) => throw new RuntimeException("JOE TBD")
-    case TypedAst.Expr.StructPut(sym, exp1, field, exp2, tpe, eff, loc) => throw new RuntimeException("JOE TBD")
-
     case TypedAst.Expr.ArrayStore(exp1, exp2, exp3, eff, loc) =>
       val e1 = visitExp(exp1)
       val e2 = visitExp(exp2)
       val e3 = visitExp(exp3)
       LoweredAst.Expr.ApplyAtomic(AtomicOp.ArrayStore, List(e1, e2, e3), Type.Unit, eff, loc)
 
+    case TypedAst.Expr.StructNew(sym, fields0, region0, tpe, eff, loc) =>
+      val fields = fields0.map{case (k, v) => (k, visitExp(v))}.sortBy{case (k, _) => k.name}
```

Comment:
And why here specifically

Reply:
The sort must occur at some point because codegen assumes that all structs with the same symbol have the same order of fields.

That said, it could occur at any point. I just figured a good place would be when we are generating the `AtomicOps`

Reply:
@mlutze @JonathanStarup Where should we do this? I would say simplifier or delay it all the way until the backend.

Reply:
There is also (now) a sort in `visitStruct`. This is because `visitType` in `simplifier` benefits from having an already sorted `Struct` object so it won't have to re-sort every time. Would it not make sense to put all the sorts together?

Reply:
Is there a reason we haven't maintained the order from parsing?

Reply:
The sorting has to happen when we convert the types (since that's also where we lose the declarations).

Also: I think this is unsafe to do naively. we can't change `new Struct{x = println(1), y = println(2)}` into `new Struct{y = println(2), x = println(1)}`. We have to let-bind the arguments in written order and then give them in sorted order.

I vote that we check that fields are given in the declaration order during frontend and then later fix that restriction

Reply:
Ah, I think you are correct. I will work on that